parlai/core/torch_agent.py

```diff
@@ -618,7 +629,12 @@ def __init__(self, opt: Opt, shared=None):
             if not shared and opt['gpu'] != -1:
                 torch.cuda.set_device(opt['gpu'])
         # indicate whether using fp16
-        self.fp16 = self.use_cuda and self.opt.get('fp16', False) and fp16_available()
+        self.fp16 = self.use_cuda and self.opt.get('fp16', False)
+        if self.fp16:
+            # check that the implementation requested is available
+            self.fp16_impl = self.opt.get('fp16_impl', 'apex')
+            if self.fp16_impl == 'apex' and not fp16_apex_available():
+                self.fp16 = False
```

Comment:
Should we print a warning in this case?

Reply:
oh yeah. we didn't before but we probably should

Reply:
actually the warning occurs in the function